[FLINK-23165][python] Add StreamExecutionEnvironment#registerSlotSharingGroup to PyFlink and Scala

SteNicholas opened the pull request:
## What is the purpose of the change

*Flink supports configuring fine grained resource requirements via DataStream API. PyFlink should add `StreamExecutionEnvironment#registerSlotSharingGroup`.*

## Brief change log

  - *Introduce `SlotSharingGroup` in pyflink. Users can use this class to describe the name and the the different resource components of a slot sharing group.*
  - *Add `DataStream#slotSharingGroup(SlotSharingGroup)`.*
  - *Add `StreamExecutionEnvironment#registerSlotSharingGroup(SlotSharingGroup)`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

KarmaGYZ commented on flink-python/pyflink/datastream/tests/test_stream_execution_environment_completeness.py:
```diff
@@ -49,7 +49,7 @@ def excluded_methods(cls):
                 'socketTextStream', 'initializeContextEnvironment', 'readTextFile',
                 'setNumberOfExecutionRetries', 'configure', 'executeAsync', 'registerJobListener',
-                'clearJobListeners', 'getJobListeners', "fromSequence", "registerSlotSharingGroup"}
+                'clearJobListeners', 'getJobListeners', "fromSequence"}
```
Also, delete the TODO above.

KarmaGYZ reviewed (commented):
Thanks for the PR, @SteNicholas . The logic generally looks good to me. We need someone who familiar with PyFlink to give another check. 
cc @HuangXingBo .

KarmaGYZ commented:
The testing fails in format checking. Would you like to fix it?

SteNicholas commented:
> The testing fails in format checking. Would you like to fix it?

I have fixed the checkstyle of the `slot_sharing_group.py`.

KarmaGYZ commented:
Just out of curiosity, is there any tool that can automatically reformat the python code?

SteNicholas commented:
> Just out of curiosity, is there any tool that can automatically reformat the python code?

No automatically reformat tool for PyFlink like Maven Spotless. Only use `lint-python.sh` to validate the checkstyle and the tests.

KarmaGYZ commented:
Maybe we can complete the missing scala interface as well in this PR, with another commit. WDYT?

SteNicholas commented:
> Maybe we can complete the missing scala interface as well in this PR, with another commit. WDYT?

@KarmaGYZ IMO, I would like to add scala interface into another commit in this PR.

KarmaGYZ commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -327,6 +327,29 @@ class DataStream[T](stream: JavaStream[T]) {
     this
   }
 
+  /**
```
It's better to be split into another commit.

HuangXingBo commented on flink-python/pyflink/datastream/slot_sharing_group.py:
```diff
@@ -0,0 +1,281 @@
+
+    def get_name(self) -> str:
+        """
+        Get the name of this SlotSharingGroup.
```
```suggestion
        Gets the name of this SlotSharingGroup.
```

KarmaGYZ replied:
May I ask why we use the singular here? Isn't it an impressive mood? Also search in Linggle, the "Get the" is used more frequently. [1][2]

[1] https://linggle.com/?q=Get+the
[2] https://linggle.com/?q=Gets+the

HuangXingBo commented on flink-python/pyflink/datastream/slot_sharing_group.py:
```diff
@@ -0,0 +1,281 @@
+
+    def get_managed_memory(self) -> Optional[MemorySize]:
+        """
+        Get the task managed memory for this SlotSharingGroup.
```
ditto

HuangXingBo commented on flink-python/pyflink/datastream/slot_sharing_group.py:
```diff
@@ -0,0 +1,281 @@
+
+    def get_task_heap_memory(self) -> Optional[MemorySize]:
+        """
+        Get the task heap memory for this SlotSharingGroup.
```
ditto

HuangXingBo commented on flink-python/pyflink/datastream/slot_sharing_group.py:
```diff
@@ -0,0 +1,281 @@
+
+        def set_cpu_cores(self, cpu_cores: float) -> 'SlotSharingGroup.Builder':
+            """
+            Set the CPU cores for this SlotSharingGroup.
```
```suggestion
            Sets the CPU cores for this SlotSharingGroup.
```

HuangXingBo reviewed (commented):
@SteNicholas Thanks a lot for the PR. Looks good overall. I only left a minor comments.

SteNicholas commented:
> @SteNicholas Thanks a lot for the PR. Looks good overall. I only left a minor comments.

@HuangXingBo , thanks for detailed review. I have addressed the comments you left. Please help to review again.

HuangXingBo reviewed (approved):
LGTM

HuangXingBo commented:
> Just out of curiosity, is there any tool that can automatically reformat the python code?

HuangXingBo commented:
@flinkbot run azure
